### PR TITLE
GG-19383 Fixed client PME hang on optimistic transactions

### PR DESCRIPTION
**Pull request by sk0x50**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheGateway.java [hunk 1]
```diff
@@ -352,7 +355,12 @@ public void onStopped() {
      * @throws IgniteException - in case of atomic operation inside transaction without permission.
      */
     private void checkAtomicOpsInTx(CacheOperationContext opCtx) throws IgniteException {
-        if (ctx.atomic() && !opCtx.allowedAtomicOpsInTx()) {
+        CacheConfiguration cfg = ctx.config();
+
+        if (cfg == null)
```
Comment (kreig133) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheGateway.java]: Why do we rely on cfg == null? Looks very fragile.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOnCachesStopTest.java [hunk 2]
```diff
@@ -254,10 +345,182 @@ private void runTxOnCacheStop(TransactionConcurrency conc, TransactionIsolation
...
+
+    /**
+     * Creates a list of cache configurations.
+     *
+     * @return List of cache configurations.
+     */
+    private List<CacheConfiguration> createCacheConfigurations() {
+        String GRP_NAME = "test-destroy-group";
+
+        List<CacheConfiguration> cacheCfgs = new ArrayList<>(50);
```
Comment (kreig133) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOnCachesStopTest.java]: What is 50? You have already CACHE_CNT, could we replace?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOnCachesStopTest.java [hunk 3]
```diff
@@ -254,10 +345,182 @@ private void runTxOnCacheStop(TransactionConcurrency conc, TransactionIsolation
...
+     * @param cacheIdxToBeDestroyed Variable that allows to get an index of destroyed cache.
+     * @param clients Client nodes that are used for initiating transactions.
+     * @param concurrency Transaction concurrency mode.
+     * @param isolation Transaction isolation.
+     * @return TxLoad future.
+     */
+    private IgniteInternalFuture startTxLoad (
+        final AtomicBoolean stopTxLoad,
+        final AtomicInteger cacheIdxToBeDestroyed,
+        final ArrayList<Ignite> clients,
```
Comment (kreig133) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOnCachesStopTest.java]: ArrayList? Maybe, just list? What do you think?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheGateway.java (hunk 1 above)
Comment (sk0x50) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheGateway.java]: Unfortunately, this method is called from `GridCacheGateway#enter(CacheOperationContext)` before the `readLock` is acquired, and therefore there is no guarantee that the cache context was not cleaned. Perhaps, it makes sense to move this check under the mentioned lock.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOnCachesStopTest.java (hunk 2 above)
Comment (sk0x50) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOnCachesStopTest.java]: Nice catch!

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOnCachesStopTest.java (hunk 3 above)
Comment (sk0x50) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOnCachesStopTest.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java [hunk 4]
```diff
@@ -3145,6 +3145,63 @@ void forceCloseCaches(ExchangeActions exchActions) {
...
     }
 
+    /**
+     * Blocks cache gateways before rolling back transactions that were affected by Exchange,
+     * which in turn was caused by the destroyCache call for these caches.
+     * This should be done in order to prohibit enlisting new entries into optimistic transactions.
+     *
+     * @param exchActions Change requests.
+     */
+    private void blockCacheGatewaysOnExchangeDone(ExchangeActions exchActions) {
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java]: This is old version of logic, also this method is not used. We should remove it.

Comment (sk0x50) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java]: Shame on me! Will remove.

**Pull request closed without merging**